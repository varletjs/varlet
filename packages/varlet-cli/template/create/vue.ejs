<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
