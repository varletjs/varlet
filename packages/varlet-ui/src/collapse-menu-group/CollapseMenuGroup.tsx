import VarIcon from '../icon'
import Ripple from '../ripple'
import { defineComponent, ref, watch, nextTick } from 'vue'
import { props } from './props'
import type { Ref } from 'vue'

import '../icon/icon'
import '../styles/common.less'
import '../ripple/ripple.less'
import './collapseMenuGroup.less'

export default defineComponent({
  name: 'VarCollapseMenuGroup',
  components: {
    VarIcon,
  },
  directives: { Ripple },
  props,
  setup(props, { slots }) {
    // TODO var-collapse-menu-item disabled 状态传递
    const contentEl: Ref<HTMLDivElement | null> = ref(null)
    const expanded: Ref<boolean> = ref(props.status === 'open')
    const show: Ref<boolean> = ref(props.status === 'open')

    // const renderTitle = () => {
    //   if (slots.title) {
    //     return slots.title()
    //   }
    //   if (isFunction(props.title)) {
    //     return props.title()
    //   }
    //   return <div>{props.title}</div>
    // }

    const handleClick = () => {
      if (props.disabled) return
      expanded.value = !expanded.value

      // TODO 完成其他逻辑
    }

    const openPanel = async () => {
      if (!contentEl.value) return
      ;(contentEl.value as HTMLDivElement).style.height = ''
      show.value = true
      await nextTick()

      const { offsetHeight } = contentEl.value as HTMLDivElement
      ;(contentEl.value as HTMLDivElement).style.height = 0 + 'px'

      requestAnimationFrame(() => {
        ;(contentEl.value as HTMLDivElement).style.height = offsetHeight + 'px'
      })
    }

    const closePanel = () => {
      if (!contentEl.value) return
      const { offsetHeight } = contentEl.value as HTMLDivElement
      ;(contentEl.value as HTMLDivElement).style.height = offsetHeight + 'px'

      requestAnimationFrame(() => {
        ;(contentEl.value as HTMLDivElement).style.height = 0 + 'px'
      })
    }

    const transitionend = () => {
      if (!expanded.value) {
        show.value = false
        ;(contentEl.value as HTMLDivElement).style.height = ''
      }
    }

    watch(expanded, (value) => {
      if (value) openPanel()
      else closePanel()
    })

    return () => {
      const prefix = 'var-collapse-menu-group'
      const { title, disabled } = props
      return (
        <div class={[prefix, { [prefix + '__disable']: disabled }]}>
          <div v-ripple={disabled} class="var-collapse-menu-group-header" onClick={handleClick}>
            <div class={prefix + '-header__title'}>{title}</div>
            <div class={prefix + '-header__icon'}>
              <var-icon
                name="chevron-down"
                transition={250}
                class={[
                  [prefix + '-header__icon'],
                  { [prefix + '-header__open']: expanded.value },
                  { [prefix + '-header__disable']: disabled },
                ]}
              />
            </div>
          </div>
          <div ref={contentEl} v-show={show.value} class={prefix + '-content'} onTransitionend={transitionend}>
            <div class={prefix + '__wrap'}>{slots.default()}</div>
          </div>
        </div>
      )
    }
  },
})
