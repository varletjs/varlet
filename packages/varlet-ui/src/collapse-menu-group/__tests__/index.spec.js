import example from '../example'
import CollapseMenuGroup from '..'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'

test('test collapse-menu-group example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
})

test('test collapse-menu-group plugin', () => {
  const app = createApp({}).use(CollapseMenuGroup)
  expect(app.component(CollapseMenuGroup.name)).toBeTruthy()
})
