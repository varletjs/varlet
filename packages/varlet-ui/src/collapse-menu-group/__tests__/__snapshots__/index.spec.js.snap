// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test collapse-menu-group example 1`] = `
"<div class=\\"collapse-demo\\">
  <div>
    <div class=\\"app-type\\">基本使用</div>
    <div class=\\"var-collapse-menu-group\\">
      <div class=\\"var-collapse-menu-group-header\\">
        <div class=\\"var-collapse-menu-group-header__title\\">标题</div>
        <div class=\\"var-collapse-menu-group-header__icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-menu-group-header__icon\\" style=\\"transition: transform 250ms;\\"></i></div>
      </div>
      <div class=\\"var-collapse-menu-group-content\\" style=\\"display: none;\\">
        <div class=\\"var-collapse-menu-group__wrap\\">
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">默认展开</div>
    <div class=\\"var-collapse-menu-group\\">
      <div class=\\"var-collapse-menu-group-header\\">
        <div class=\\"var-collapse-menu-group-header__title\\">标题</div>
        <div class=\\"var-collapse-menu-group-header__icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-menu-group-header__icon var-collapse-menu-group-header__open\\" style=\\"transition: transform 250ms;\\"></i></div>
      </div>
      <div class=\\"var-collapse-menu-group-content\\">
        <div class=\\"var-collapse-menu-group__wrap\\">
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div class=\\"app-type\\">禁用</div><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--2\\" style=\\"margin-bottom: 8px;\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">禁用</div>
    </button>
    <div class=\\"var-collapse-menu-group\\">
      <div class=\\"var-collapse-menu-group-header\\">
        <div class=\\"var-collapse-menu-group-header__title\\">标题</div>
        <div class=\\"var-collapse-menu-group-header__icon\\"><i class=\\"var-icon var-icon--set var-icon-chevron-down var-collapse-menu-group-header__icon\\" style=\\"transition: transform 250ms;\\"></i></div>
      </div>
      <div class=\\"var-collapse-menu-group-content\\" style=\\"display: none;\\">
        <div class=\\"var-collapse-menu-group__wrap\\">
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
          <div class=\\"var-collapse-menu-item\\">
            <div class=\\"var-collapse-menu-item__indent\\" style=\\"padding-left: 20px;\\"></div>
            <div class=\\"var-collapse-menu-item__icon\\">
              <!---->
            </div>
            <div class=\\"var-collapse-menu-item__label\\">
              <div>文本</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
